src/Features/ExternalAccess/Copilot/Internal/Completion/CSharpContextProviderService.cs

```diff
@@ -6,80 +6,34 @@
-    {
-        Providers = providers.ToImmutableArray();
-    }
-
-    public async IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, [EnumeratorCancellation] CancellationToken cancellationToken)
-    {
-        var queue = new AsyncQueue<IContextItem>();
-        var tasks = this.Providers.Select(provider => Task.Run(async () =>
-        {
-            try
-            {
-                await provider.ProvideContextItemsAsync(document, position, activeExperiments, ProvideItemsAsync, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception exception) when (FatalError.ReportAndCatchUnlessCanceled(exception, ErrorSeverity.General))
-            {
-            }
-        },
-        cancellationToken));
-
-        // Let all providers run in parallel in the background, so we can steam results as they come in.
-        // Complete the queue when all providers are done.
-        _ = Task.WhenAll(tasks)
-            .ContinueWith((_, __) => queue.Complete(),
-                          null,
-                          cancellationToken,
-                          TaskContinuationOptions.ExecuteSynchronously,
-                          TaskScheduler.Default);
-
-        while (true)
-        {
-            IContextItem item;
-            try
-            {
-                item = await queue.DequeueAsync(cancellationToken).ConfigureAwait(false);
-            }
-            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
-            {
-                // Dequeue is cancelled because the queue is empty and completed, we can break out of the loop.
-                break;
-            }
-
-            yield return item;
-        }
-
-        ValueTask ProvideItemsAsync(ImmutableArray<IContextItem> items, CancellationToken cancellationToken)
-        {
-            foreach (var item in items)
-            {
-                queue.Enqueue(item);
-            }
-
-            return default;
-        }
-    }
+    private readonly ImmutableArray<IContextProvider> _providers = [.. providers];
+
+    public IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, CancellationToken cancellationToken)
+        => ProducerConsumer<IContextItem>.RunParallelStreamAsync(
+            _providers,
+            static async (provider, callback, args, cancellationToken) =>
+                await provider.ProvideContextItemsAsync(
+                    args.document, args.position, args.activeExperiments,
+                    (items, cancellationToken) =>
+                    {
+                        foreach (var item in items)
+                            callback(item);
+
+                        return default;
+                    }, cancellationToken).ConfigureAwait(false),
```

Comment:
note: if ProvideContextItemsAsync just took an Action<ICallbackItem> this could would be simplified to:

```suggestion
                    (provider, cancellationToken) => provider.ProvideContextItemsAsync(
                        document, position, activeExperiments, callback, cancellationToken);
```

Which is much nicer and cleaner. It's unclear why we're passing a callback that has to package up everything in an array and be async.